### PR TITLE
Controller changes to enable modular upgrades for tinkerbell provider

### PR DESCRIPTION
**Pull request by rahulbabu95**
*Issue #, if available:*

*Description of changes:*
Controller changes to support modular upgrades for Tinkerbell provider. The change also introduces a check to verify the Kubernetes version is specified on the OSImage URL field.  

*Testing (if applicable):*
Tested the FLC changes manually on workload cluster for both CP and worker upgrades separately.

*Documentation added/planned (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by chrisdoherty4 — COMMENTED** [review]
Rahul and I talked through a simplification of the validation function. See https://github.com/rahulbabu95/eks-anywhere/pull/8/files.

tl;dr the simplification supports the case where a single image is used and reduces the number of functions making it easier to read for maintainers.

File: pkg/providers/tinkerbell/reconciler/reconciler.go [hunk 1]
```diff
@@ -182,6 +182,16 @@ func (r *Reconciler) DetectOperation(ctx context.Context, log logr.Logger, tinke
 		return K8sVersionUpgradeOperation, nil
 	}
 
+	for _, wg := range tinkerbellScope.Workers.Groups {
+		machineDeployment, err := controller.GetMachineDeployment(ctx, r.client, wg.MachineDeployment.GetName())
+		if err != nil {
+			return "", errors.Wrap(err, "failed to get workernode group machinedeployment")
+		}
+		if machineDeployment != nil && (*machineDeployment.Spec.Template.Spec.Version != *wg.MachineDeployment.Spec.Template.Spec.Version) {
+			log.Info("Operation detected", "operation", K8sVersionUpgradeOperation)
```
[the remark this answers is not in the document]
Comment (rahulbabu95) [reply pkg/providers/tinkerbell/reconciler/reconciler.go:191]: From the controller logs, its pretty hard to follow the current operation and it could be make it easier to validate about the current reconciliation with this log line.I personally experienced this while manually testing out modular upgrades!

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go [hunk 2]
```diff
@@ -100,17 +100,8 @@ func GetDefaultActionsFromBundle(clusterSpec *Cluster, b v1alpha1.VersionsBundle
...
-		switch {
-		case osImageOverride != "":
-			imageURL = osImageOverride
-		default:
-			imageURL = b.EksD.Raw.Bottlerocket.URI
-		}
-
 		env := map[string]string{
 			"DEST_DISK":  disk,
 			"IMG_URL":    imageURL,
```
[the remark this answers is not in the document]
Comment (rahulbabu95) [reply pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go:107]: Yes, we set it at the machine config level during the validation for BR

File: pkg/providers/tinkerbell/validate.go [hunk 3]
```diff
@@ -48,6 +49,64 @@ func validateOSImageURL(spec *ClusterSpec) error {
...
+		if err := checkK8sVersionMatch(k8sVersion, osImageURL); err != nil {
+			return fmt.Errorf("kubernetes version check in OSImageURL failed for %s: %v", wngRefName, err)
+		}
+	}
+	return nil
+}
+
+func checkK8sVersionMatch(k8sVersion, imageURL string) error {
+	versionExtractor := strings.NewReplacer("-", "", ".", "", "_", "")
+	OSimageURL := versionExtractor.Replace(imageURL)
```
[the remark this answers is not in the document]
Comment (rahulbabu95) [reply pkg/providers/tinkerbell/validate.go:119]: umm., for now let's keep this similar across provider and can get back later to this if we want to change it to a regex and place it at a level higher that could be reused!
